### DISCUSSION TITLE
string[] to byte ** / sbyte  ** and byte ** / sbyte  ** to string[] for example?

### DISCUSSION THREAD

**Discussion by DeafMan1983** [dotnet/runtime#85740] [Q&A] [answered]
Hello everyone, 

I would like to know how do I convert `sbyte *` to `string`  and `string` to `sbyte *` but I have made in DeafMan1983.Conversion and but I want know how do I `string[]` to `sbyte **` and `sbyte **` to `string[]` and I don't get already to find Google but I don't have idea. I tried to write string[] to sbyte **. No success...

Do you have idea with double pointers to string[] args

Because sometimes `UnmanagedCallerOnly` doesn't like to pass string[] if parameter of unmanaged function like example
```
[UnmanagedCallerOnly(EntryPoint = "MyFunc")]
Public void MyFunc(sbyte **args)
{
    // pass to string[] with for statement
    string[] sArgs = StringArrayFromCharDoublePointer(args);
    for (int i = 0; i < sArgs.Length: i++)
    {
        // pass in command line like this sArgs[i] = "--help" etc
    }
}
```

Program pass with `delegate *unmanaged[Cdecl]<sbyte **, void> pfnMyFunc` with `dlsym` sorry I use only Linux not Windows.

Example with `Main(string[] args)`
```
static void Main(string[] args)
{
   ....
   delegate *unmanaged[Cdecl]<sbyte **, void> pfnMyFunc = (delegate *unmanaged[Cdecl]<sbyte **, void>)dlsym(CharPtrsfromString("MyFunc"));
    // that is passing args with external functions from native library 
    pfnMyFunc(CharDoublePtrsFromStringArray(args));
}
```

A Method `sbyte **CharDoublePtrsFromStringArray(string[] args)` is for double pointers to string array and other method `string[] StringArrayFromCharDoublePointer(sbyte **unmanagedargs)`

I have problem with multiple arrays or pointers. How do I know like this? Can you help me if you have ideas?

**Comment by jkotas** [comment]
Here is an example where we do something similar in this repo: https://github.com/dotnet/runtime/blob/cb5fe56312c40244524f2bd5d89006279e52d1da/src/libraries/Common/src/Interop/Unix/System.Native/Interop.ForkAndExecProcess.cs#L50-L89 .

The idea is to allocate the converted strings in native memory. It is necessary to free them once you are done with it.

**Reply by DeafMan1983** [reply]
Uh uh it looks weird for void and in. It is not way to return of string[] and sbyte ** or byte **

I believe that way from string[] to sbyte ** 1. foreach, 2.for i++ 3 fixed to sbyte **.
Way from sbyte ** to string[] 1. for i++ 2. foreach return outside of () of for i++ 3 return stringarray; So that I believe my methods are better but I don't test it and I will try later. 

Now I try to resolve but still problem `NullReferenceException` :(

Check out example my idea:
```
    public static sbyte **SByteDoublePointersWithStringArray(string[] arrays)
    {
        sbyte **array_ptrs = null;
        for (int i = 0; i < arrays.Length; i++)
        {
            array_ptrs[i] = SBytePointerWithString(arrays[i]);
        }

        return array_ptrs;
    }

    public static string[] StringArrayWithSByteDoublePointers(sbyte **array_ptrs)
    {
        Span<string> span_arrays = new Span<string>();
        for (int i = 0; i < span_arrays.Length; i++)
        {
            string[] arrays = span_arrays.ToArray();;
            arrays[i] = StringwithSBytePointer(array_ptrs[i]);
        }

        return span_arrays.ToArray();;
    }
```
Check [DeafMan1983.Conversion](https://github.com/DeafMan1983/DeafMan1983.Conversion)

But It seems failed for sbyte ** to string[] but I don't use Memory. just normal array of string like from `static void Main(string[] args)`
And `delegate unmanaged[Cdecl]<******>` if it catches sbyte ** from string[] under external native library like this
`./GameLauncher --help`
`--help` is from Engine in `sArgs[i]` with `sbyte ** args`

It is hard to resolve.....

**Reply by DeafMan1983** [reply]
Hello I have another methods but it throws NullReferenceException problem :(

Engine ( native library )

Host.cs
```
    [UnmanagedCallersOnly(EntryPoint = "Host_Main")]
    public static int Host_Main(sbyte** args, sbyte* progname, int bchange_name, delegate* <string, void> func)
    {
        Console.WriteLine("Hello Half-Life Engine, lol!");
        string[] sArgs = Environment.GetCommandLineArgs();

        pChangeGame = func;

        if (!Directory.Exists(StringwithSBytePointer(progname)))
        {
            Console.WriteLine("Not found, Please re-update Steam!");
        }
        else
        {
            Console.WriteLine("Woohoo, Half-Life Sharp (x64)");
            Console.WriteLine("All executables and libraries are written in C#\n" +
            "and are compiled in NativeAot with Shared. :-D");
        }

        for (int i = 0; i < sArgs.Length; i++)
        {
            args[i] = SBytePointerWithString(sArgs[i]);
            if (args[i] == SBytePointerWithString("help"))
            {
                Console.WriteLine("Display help!");
            }
            if (sArgs[i] == "quit")
            {
                Environment.Exit(0);
            }
        }

        return 0;
    }
```

And GameLauncher:
Program.cs

```
    ....
    public static delegate *<sbyte **, sbyte *, int, delegate* unmanaged[Cdecl]<sbyte *, void>, int> pfnMain_p_new;
    ....
            pfnMain_p_new = (delegate *<sbyte **, sbyte *, int, delegate* unmanaged[Cdecl]<sbyte *, void>, int> )dlsym(hEngine, SBytePointerWithString("Host_Main"));
            if (pfnMain_p_new == null)
            {
                Console.WriteLine("Unable to pass function name {0}", GetError());
            }
    ....
    public static int Sys_Start()
    {
        Sys_LoadEngine();
        delegate* unmanaged[Cdecl]<sbyte *, void> pfnChangeGame = &Sys_ChangeGame;
        sbyte *arg = SBytePointerWithString(string.Join(" ", szArgs));
        sbyte **args = stackalloc sbyte*[]{ arg };
        int returnCode = pfnMain_p_new(args, SBytePointerWithString(GAMEDIR), 0, pfnChangeGame);
        Sys_UnloadEngine();
        return returnCode;
    }

    static int Main(string[] args)
    {
        return Sys_Start();
    }
```

And I tested in commandline ./GameLauncher if not parameter then I use "help" as parameter ./GameLauncher help
Result:
![image](https://user-images.githubusercontent.com/57066679/236238916-78ab1a99-0fdf-4492-b013-b3cd37d11587.png)

I tried and get nothing work..... :(

**Reply by jkotas** [reply]
Are you missing `umanaged` on `delegate* <string, void> func`?

**Reply by DeafMan1983** [reply]
Thanks for fix but still nothing display Console from external native library :/

Fullcode:
Program.cs
```
namespace valve;

using static DeafMan1983.ConvFunctions;

using System.IO;
using System.Runtime.InteropServices;
using System.Runtime.CompilerServices;

unsafe class Game
{
    // For Linux:
    private const string dl = "c";    
    
    [DllImport(dl)]
    private static extern void *dlopen(sbyte *so_file, uint flags);

    [DllImport(dl)]
    private static extern void *dlsym(void *so_handle, sbyte *func_name);

    [DllImport(dl)]
    private static extern void dlclose(void *so_handle);

    [DllImport(dl)]
    private static extern sbyte *dlerror();
    public static delegate *unmanaged[Cdecl]<sbyte **, int, sbyte *, int, delegate* unmanaged[Cdecl]<sbyte *, void>, int> pfnMain_p_new;
    public static delegate* unmanaged[Cdecl]<void> pfnShutdown_p_new;

    private static string szGameDir = string.Empty;
    private static string[] szArgs = Environment.GetCommandLineArgs();

    private static string GAMEDIR = Path.Combine(AppDomain.CurrentDomain.BaseDirectory, "valve");

    private static string GetError()
    {
        return StringwithSBytePointer(dlerror());
    }
    private static void *hEngine;

    private static void FreeLibrary()
    {
        if (hEngine != null)
        {
            dlclose(hEngine);
        }
    }

    private static string strncpy(ref string dst, string src, int num)
    {
        return dst=src.Substring(0,num);
    }


    private static void Sys_LoadEngine()
    {
        if (RuntimeInformation.IsOSPlatform(OSPlatform.Linux))
        {
            hEngine = dlopen(SBytePointerWithString(Path.Combine(AppDomain.CurrentDomain.BaseDirectory, "Engine.so")), 0x00002);
            if (hEngine == null)
            {
                Console.WriteLine("Unable to load {0}", GetError());
            }

            pfnMain_p_new = (delegate *unmanaged[Cdecl]<sbyte **, int, sbyte *, int, delegate* unmanaged[Cdecl]<sbyte *, void>, int> )dlsym(hEngine, SBytePointerWithString("Host_Main"));
            if (pfnMain_p_new == null)
            {
                Console.WriteLine("Unable to pass function name {0}", GetError());
            }
        }

        pfnShutdown_p_new = (delegate *unmanaged[Cdecl]<void>)dlsym(hEngine, SBytePointerWithString("Host_Shutdown"));
    }

    private static void Sys_UnloadEngine()
    {
        if (pfnShutdown_p_new != null)
        {
            pfnShutdown_p_new();
        }

        if (hEngine != null)
        {
            FreeLibrary();
        }

        pfnMain_p_new = null;
	    pfnShutdown_p_new = null;
    }

    private static sbyte **GetIndex(string[] args)
    {
        sbyte **sArgs = stackalloc sbyte*[args.Length];
        for (int i = 0; i < args.Length; i++)
        {
            sArgs[i] = SBytePointerWithString(args[i]);
        }
        return sArgs;
    }

    [UnmanagedCallersOnly(CallConvs = new[] {typeof(CallConvCdecl)})]
    private static void Sys_ChangeGame(sbyte *progname)
    {
        if (progname == null)
        {
            Console.Error.WriteLine("Error: Sys_ChangeGame is null.");
        }

        if (pfnShutdown_p_new == null)
        {
            Console.Error.WriteLine("Error: Sys_ChangeGame: missed 'Host_Shutdown'.");
        }

        strncpy(ref szGameDir, StringwithSBytePointer(progname), szGameDir.Length - 1);
        Sys_UnloadEngine();
        Sys_LoadEngine();
        delegate* unmanaged[Cdecl]<sbyte *, void> pfnChangeGame = &Sys_ChangeGame;
        pfnMain_p_new(GetIndex(szArgs), szArgs.Length, SBytePointerWithString(szGameDir), 1, pfnChangeGame);
    }

    public static int Sys_Start()
    {
        Sys_LoadEngine();
        delegate* unmanaged[Cdecl]<sbyte *, void> pfnChangeGame = &Sys_ChangeGame;
        int returnCode = pfnMain_p_new(GetIndex(szArgs), szArgs.Length, SBytePointerWithString(GAMEDIR), 0, pfnChangeGame);
        Sys_UnloadEngine();
        return returnCode;
    }

    static int Main(string[] args)
    {
        return Sys_Start();
    }
}
```
for Launcher

and Host.cs from Engine, classlib as native library
```
namespace Engine;

using System;
using System.IO;
using System.Runtime.InteropServices;

using static DeafMan1983.ConvFunctions;

public unsafe class Host
{
    private static delegate* unmanaged[Cdecl]<string, void> pChangeGame = null;


    // ...

    private static string[] FromIndex(sbyte ** sArrays, int array_length)
    {
        string[] arrays = new string[array_length];
        for (int i = 0; i < array_length; i++)
        {
            arrays[i] = StringwithSBytePointer(sArrays[i]);
        }
        return arrays;
    }

    // ------------------------------------------------
    // Host_Main()
    // ------------------------------------------------
    [UnmanagedCallersOnly(EntryPoint = "Host_Main")]
    public static int Host_Main(sbyte** args, int array_length, sbyte* progname, int bchange_name, delegate* unmanaged[Cdecl]<string, void> func)
    {
        Console.WriteLine("Hello Half-Life Engine, lol!");

        pChangeGame = func;

        if (!Directory.Exists(StringwithSBytePointer(progname)))
        {
            Console.WriteLine("Not found, Please re-update Steam!");
        }
        else
        {
            Console.WriteLine("Woohoo, Half-Life Sharp (x64)");
            Console.WriteLine("All executables and libraries are written in C#\n" +
            "and are compiled in NativeAot with Shared. :-D");
        }

        string[] sArgs = FromIndex(args, array_length);
        for (int i = 0; i < sArgs.Length; i++)
        {
            // Parammeter "help" example: ./Launcher help
            if (sArgs[i] == "help")
            {
                Console.WriteLine("Display help!");
            }
        }

        return 0;
    }

    // ------------------------------------------------
    // Host_Shutdown()
    // ------------------------------------------------
    [UnmanagedCallersOnly(EntryPoint = "Host_Shutdown")]
    public static void Host_Shutdown()
    {
        Environment.Exit(0);
    }
}
```

And result looks without argument `help`
![image](https://user-images.githubusercontent.com/57066679/236267486-fddeb696-bb42-43bf-bd62-cb00a555f9ea.png)

And I try to pass argument `help`
![image](https://user-images.githubusercontent.com/57066679/236267613-e68362c8-a98d-4e63-ad33-a1cc7a384314.png)
Eh where is console line? Do you think I need to exercise AppWithPlugins Example into Native Aot?

**Reply by jkotas** [reply]
```
    private static sbyte **GetIndex(string[] args)
    {
        sbyte **sArgs = stackalloc sbyte*[args.Length];
        for (int i = 0; i < args.Length; i++)
        {
            sArgs[i] = SBytePointerWithString(args[i]);
        }
        return sArgs;
    }
```
This is wrong. You cannot return stackallocated memory from a function that stackallocated it.

**Reply by DeafMan1983** [reply]
Then ???

I have tried.....
```
    private static sbyte **GetIndex(string[] args)
    {
        sbyte[][] sArgs = new sbyte[args.Length][];
        for (int i = 0; i < args.Length; i++)
        {
            fixed (sbyte** sArgsPtrs = sArgs[i][i])
            {
                sArgsPtrs[i] = SBytePointerWithString(args[i]);
                return sArgsPtrs;
            }
        }
    }
```
sbyte* =/= sbyte[]
sbyte** =/= sbyte[][]

How do I get successfully?

**Reply by jkotas** [reply, marked as answer]
Either allocate array of pointers `sbyte*[] sArgs = new sbyte*[args.Length];` and pin it using fixed; 
or allocate native memory using `sbyte** sArgs = (sbyte**)NativeMemory.Alloc(args.Length * sizeof(sbyte*));` and then free it using `NativeMemory.Free` once you are done with it.

**Reply by DeafMan1983** [reply]
Thanks for Fixed wow why do I need NativeMemory? I expect Memory Leak. Thanks for supporting me. I need invite you for rewards  😋🙂

**Reply by DeafMan1983** [reply]
Hey my friend,  person from discord told me because `NativeMemory()` has leak. I don't understand why does it happen?
Proof screenshot of discord:
[Check out](https://i.imgur.com/8GqWdhw.jpeg).

**Reply by stephentoub** [reply]
What is your code?

**Reply by DeafMan1983** [reply]
Check out my repository [code](https://github.com/DeafMan1983/DeafMan1983.Conversion/blob/d41eef48c45d1e5e6634a75a7caa1440c6f80052/ConvFunctions.cs#L77). Has it Memory Leak? But I already tested with my own game with loading native library engine.  And it has not problem. Who said because my Conversion for `sbyte **` and `string[]`?

**Reply by stephentoub** [reply]
You're calling NativeMemory.Alloc. Where are you calling NativeNemory.Free?

**Reply by DeafMan1983** [reply]
Why do I need it? But how do I pass return type when I use NativeMemory.Free() then it won't return with return type. Am I correct or wrongly? I will check out later if I am after my work.

**Reply by jkotas** [reply]
> Why do I need it? 

Memory allocated using `NativeMemory.Alloc` is same as memory allocated using `malloc` in C. This memory is not managed by the garbage collector. You have to explicitly call `NativeMemory.Free` once it is no longer needed. You need to add a new method to your ConvFunctions class to free the memory. The code that uses ConvFunctions should call this new method once the converted array is no longer needed. 

When you are using `NativeMemory.Alloc` you have to follow same memory management rules as if you were writing C code.

**Reply by DeafMan1983** [reply]
I have found website but only IntPtr =/= sbyte ** to string[] is it correct or wrong for StringWithSByteDoublePointer() without memory leak? Like while statement? Example [click here](https://www.appsloveworld.com/csharp/100/287/how-to-convert-the-intptr-to-an-array)! Please remember that cause Marshal class cannot work for NativeAot! That's why I tried to resolve problems with memory leaks. My god That's hard to understand but I am learning managed pointer to string array or string array to managed pointer.

But NativeMemory.free() after Alloc is really impossible if you are ready then you cannot get line like I explain you 

NativeAot end used format without Marshal.cs 
```
Executable.                   Native library 
string[] to sbyte ** -------- sbyte ** to string[]
args[i] finds "help"  < -- >  "help" from string[i]
```

That is problematic cause delegate *unmanaged[Cdecl] doesn't like to use string[] if I use `UnmanagedCallerOnly` from Native Library.
